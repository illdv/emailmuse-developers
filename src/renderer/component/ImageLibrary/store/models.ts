export interface IFileInfo {
  id: number;
  name: string;
  hash: string;
  imagable_id: number;
  imagable_type: string;
  created_at: string;
  updated_at: string;
  deleted_at?: string;
  url: string;
  thumb_url: string;
}

export interface IGetImagesResponse {
  current_page: number;
  data: IFileInfo[];
  first_page_url: string;
  from: number;
  last_page: number;
  last_page_url: string;
  next_page_url: null;
  path: string;
  per_page: number;
  prev_page_url: string;
  to: number;
  total: number;
}

export interface IImageLibraryItem {
  id: number;
  /**
   * Original image.
   */
  url: string;
  /**
   * Use for preview image.
   */
  thumb_url: string;
  name: string;
}