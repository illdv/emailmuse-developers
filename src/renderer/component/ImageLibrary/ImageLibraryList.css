.image-library-list {
  width: 100%;
  height: 100%;
}

.image-library-list__tile {
  user-select: none;
}

.image-library-list__tile-img {
  cursor: pointer;
}