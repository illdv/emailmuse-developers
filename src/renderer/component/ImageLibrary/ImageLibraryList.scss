.image-library-list {
  width: 100%;
  height: 100%;
}

.image-library-list__tile {
  user-select: none;
}

.image-library-list__tile-img {
  width: 100%;
  min-height: 300px;
  cursor: pointer;
}