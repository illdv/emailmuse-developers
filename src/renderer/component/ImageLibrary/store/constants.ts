export const GET_IMAGES_REQUEST = 'GET_IMAGE_REQUEST';
export const GET_IMAGES_SUCCESS = 'GET_IMAGE_SUCCESS';
export const GET_IMAGES_FAILURE = 'GET_IMAGE_FAILURE';

export const UPLOAD_IMAGES_REQUEST = 'UPLOAD_IMAGE_REQUEST';
export const UPLOAD_IMAGES_SUCCESS = 'UPLOAD_IMAGE_SUCCESS';
export const UPLOAD_IMAGES_FAILURE = 'UPLOAD_IMAGE_FAILURE';

export const UPDATE_IMAGE_REQUEST = 'UPDATE_IMAGE_REQUEST';
export const UPDATE_IMAGE_SUCCESS = 'UPDATE_IMAGE_SUCCESS';
export const UPDATE_IMAGE_FAILURE = 'UPDATE_IMAGE_FAILURE';

export const DELETE_IMAGES_REQUEST = 'DELETE_IMAGE_REQUEST';
export const DELETE_IMAGES_SUCCESS = 'DELETE_IMAGE_SUCCESS';
export const DELETE_IMAGES_FAILURE = 'DELETE_IMAGE_FAILURE';