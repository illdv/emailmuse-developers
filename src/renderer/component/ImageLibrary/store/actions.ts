import * as constants from './constants';
import { createAction } from 'redux-actions';
import { IGetImagesResponse } from 'src/renderer/component/ImageLibrary/store/models';
import * as Status from 'src/renderer/common/PreloaderLayout/Status/constants';

export const getImagesRequest = (page = 1, name: string = '') => ({
  type: constants.GET_IMAGES_REQUEST,
  payload: { page, name},
  meta: { status: Status.LOADING },
});

export const getImagesSuccess = (response: IGetImagesResponse) => ({
  type: constants.GET_IMAGES_SUCCESS,
  payload: response,
  meta: { status: Status.LOADED },
});

export const getImagesFailure = createAction(constants.GET_IMAGES_FAILURE, null, () => ({ status: Status.FAILED }));

export const uploadImagesRequest = (files: File | File[]) => ({
  type: constants.UPLOAD_IMAGES_REQUEST,
  payload: [].concat(files),
  meta: { status: Status.LOADING },
});

export const uploadImagesSuccess =
  createAction(constants.UPLOAD_IMAGES_SUCCESS, null, () => ({ status: Status.LOADED }));

export const uploadImagesFailure =
  createAction(constants.UPLOAD_IMAGES_FAILURE, null, () => ({ status: Status.FAILED }));

export const updateImageRequest = (payload: { imageId: number, name: string }) => ({
  type: constants.UPDATE_IMAGE_REQUEST,
  payload,
  meta: { status: Status.LOADING },
});

export const updateImageSuccess =
  createAction(constants.UPDATE_IMAGE_SUCCESS, null, () => ({ status: Status.LOADED }));

export const updateImageFailure =
  createAction(constants.UPDATE_IMAGE_FAILURE, null, () => ({ status: Status.FAILED }));

export const deleteImagesRequest = (ids: number | number[]) => ({
  type: constants.DELETE_IMAGES_REQUEST,
  payload: [].concat(ids),
  meta: { status: Status.LOADING },
});

export const deleteImagesSuccess =
  createAction(constants.DELETE_IMAGES_SUCCESS, null, () => ({ status: Status.LOADED }));

export const deleteImagesFailure =
  createAction(constants.DELETE_IMAGES_FAILURE, null, () => ({ status: Status.FAILED }));
