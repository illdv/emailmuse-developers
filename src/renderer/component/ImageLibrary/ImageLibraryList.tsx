import * as React from 'react';
import { GridList, GridListTile, GridListTileBar, Grow, IconButton } from '@material-ui/core';
import block from 'bem-ts';

import { IImageLibraryItem } from 'src/renderer/component/ImageLibrary/store/models';
import 'src/renderer/component/ImageLibrary/ImageLibraryList.scss';
import { Delete } from '@material-ui/icons';

const b = block('image-library-list');

namespace ImageLibraryListSpace {
  export interface IProps {
    items: IImageLibraryItem[];
    onDelete?: (item: IImageLibraryItem) => () => void;
    onSelect: (item: IImageLibraryItem) => () => void;
  }

  export interface IState {

  }
}

export class ImageLibraryList extends React.Component<ImageLibraryListSpace.IProps, ImageLibraryListSpace.IState> {

  constructor(props) {
    super(props);
  }

  render() {
    const { onDelete, onSelect, items } = this.props;
    return (
      <>
        <GridList
          cellHeight={240}
          cols={3}
          className={b()}
        >
          {
            items.map((item, index) => (
                <Grow key={item.id} in timeout={(index * 500) + 500}>
                  <GridListTile className={b('tile')}>
                    <img
                      className={b('tile-img')}
                      src={item.thumb_url}
                      onClick={onSelect(item)}
                    />
                    {
                      onDelete &&
                      <GridListTileBar
                        title={item.name}
                        actionIcon={
                          <IconButton onClick={onDelete(item)}>
                            <Delete nativeColor='white'/>
                          </IconButton>
                        }
                      />
                    }
                  </GridListTile>
                </Grow>
              ),
            )
          }
        </GridList>
      </>
    );
  }
}
