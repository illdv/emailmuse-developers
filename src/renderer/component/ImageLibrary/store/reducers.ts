import { IFileInfo, IPagination }from 'src/renderer/component/ImageLibrary/store/models';
import { handleActions } from 'redux-actions';
import * as constants from 'src/renderer/component/ImageLibrary/store/constants';

export namespace ImageLibrary {
  export interface IState {
    all: IFileInfo[];
    pagination: IPagination;
  }

  const createDefaultState = (): IState => {
    return {
      all: [],
      pagination: {
        current_page: null,
        last_page: null,
        per_page: null,
        total: null
      },
    };
  };

  export const reducer = handleActions(
    {
      [constants.GET_IMAGES_SUCCESS]: (state: IState, action): IState => {
        return { ...state, all: action.payload.data,
          pagination: {
            current_page: action.payload.current_page,
            last_page: action.payload.last_page,
            per_page: action.payload.per_page,
            total: action.payload.total,
          }
        };
      },
    },
    createDefaultState()
  );
}