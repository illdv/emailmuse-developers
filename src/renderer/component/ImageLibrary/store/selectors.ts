import { createSelector } from 'reselect';
import { IFileInfo, IImageLibraryItem } from 'src/renderer/component/ImageLibrary/store/models';

const imagesSelector = (state):IFileInfo[] => state.images.all;

export const getImagesURLSelector = createSelector(
  imagesSelector,
  (state: IFileInfo[]): IImageLibraryItem[] =>
    state.map(file => {
      return {
        id: file.id,
        url: file.url,
        name: file.name
      };
    })
);
