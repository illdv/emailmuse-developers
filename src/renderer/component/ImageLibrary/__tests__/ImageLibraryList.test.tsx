import * as React from 'react';
import {compose} from 'redux';
import {ImageLibraryList} from '../ImageLibraryList';
import {mount, shallow} from 'enzyme';
import {shallowToJson} from 'enzyme-to-json';
import {IImageLibraryItem} from '../store/models';

// jest.mock('../../../src/components/ui/Color', () => {
//     ({rating, onRate = f => f, onRemove = f => f}) => {}
// });

describe('<ImageLibraryList /> UI Component', () => {
    const image0: IImageLibraryItem = {
        id: 89,
        url: 'https://infocron.s3.us-east-2.amazonaws.com/user-5/nc6DTVcBCpU3Vnew8QuoZkklwvN2uWukO8fPfrSg.jpeg',
        thumb_url: 'https://infocron.s3.us-east-2.amazonaws.com/' +
        'user-5/nc6DTVcBCpU3Vnew8QuoZkklwvN2uWukO8fPfrSg_thumb450.jpeg',
        name: 'tesla-cat',
    };
    const image1: IImageLibraryItem = {
        id: 90,
        url: 'https://infocron.s3.us-east-2.amazonaws.com/user-5/nc6DTVcBCpU3Vnew8QuoZkklwvN2uWukO8fPfrSg.jpeg',
        thumb_url: 'https://infocron.s3.us-east-2.amazonaws.com/' +
        'user-5/nc6DTVcBCpU3Vnew8QuoZkklwvN2uWukO8fPfrSg_thumb450.jpeg',
        name: 'tesla-cat',
    };
    const image2: IImageLibraryItem = {
        id: 91,
        url: 'https://infocron.s3.us-east-2.amazonaws.com/user-5/nc6DTVcBCpU3Vnew8QuoZkklwvN2uWukO8fPfrSg.jpeg',
        thumb_url: 'https://infocron.s3.us-east-2.amazonaws.com/' +
        'user-5/nc6DTVcBCpU3Vnew8QuoZkklwvN2uWukO8fPfrSg_thumb450.jpeg',
        name: 'tesla-cat',
    };
    const imageArray: IImageLibraryItem[] = [image0, image1, image2];
    const selectImg = jest.fn();
    const deleteImg = jest.fn();
    let props = {
        items: [...imageArray],
        onDelete: deleteImg,
        onSelect: selectImg,
    };
    const imageList = mount(<ImageLibraryList {...props} />);

    describe('Rendering UI', () => {
        it('Renders Correctly', () => {
            compose(expect, shallowToJson, shallow)(<ImageLibraryList {...props} />).toMatchSnapshot();
        });

        it('Defaults Properties correctly', () => {
            props = {
                items: [],
                onDelete: jest.fn(),
                onSelect: jest.fn(),
            };
            expect(shallow(<ImageLibraryList {...props} />));
        });

        describe('Selecting a img', () => {
            imageList.find('img')
                .first()
                .simulate('click');

            it('must dispatch onSelect', () => {
                expect(selectImg).toBeCalled();
            });

            it('onSelect take image', () => {
                expect(selectImg).toBeCalledWith(image0);
            });

        });

        describe('Clicking delete button', () => {
            imageList
                .find('IconButton')
                .first()
                .simulate('click');

            it('must dispatch onDelete', () => {
                expect(deleteImg).toBeCalled();
            });

            it('onDelete take image', () => {
                expect(deleteImg).toBeCalledWith(image0);
            });

        });
    });
});
