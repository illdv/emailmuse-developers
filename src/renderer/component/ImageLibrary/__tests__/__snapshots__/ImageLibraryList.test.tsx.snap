// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ImageLibraryList /> UI Component Rendering UI Renders Correctly 1`] = `
<React.Fragment>
  <WithStyles(GridList)
    cellHeight={240}
    className="image-library-list"
    cols={3}
  >
    <WithTheme(Grow)
      in={true}
      key="89"
      timeout={500}
    >
      <WithStyles(GridListTile)
        className="image-library-list__tile"
      >
        <img
          className="image-library-list__tile-img"
          src="https://infocron.s3.us-east-2.amazonaws.com/user-5/nc6DTVcBCpU3Vnew8QuoZkklwvN2uWukO8fPfrSg_thumb450.jpeg"
        />
        <WithStyles(GridListTileBar)
          actionIcon={
            <WithStyles(IconButton)
              onClick={undefined}
            >
              <pure(Delete)
                nativeColor="white"
              />
            </WithStyles(IconButton)>
          }
          title="tesla-cat"
        />
      </WithStyles(GridListTile)>
    </WithTheme(Grow)>
    <WithTheme(Grow)
      in={true}
      key="90"
      timeout={1000}
    >
      <WithStyles(GridListTile)
        className="image-library-list__tile"
      >
        <img
          className="image-library-list__tile-img"
          src="https://infocron.s3.us-east-2.amazonaws.com/user-5/nc6DTVcBCpU3Vnew8QuoZkklwvN2uWukO8fPfrSg_thumb450.jpeg"
        />
        <WithStyles(GridListTileBar)
          actionIcon={
            <WithStyles(IconButton)
              onClick={undefined}
            >
              <pure(Delete)
                nativeColor="white"
              />
            </WithStyles(IconButton)>
          }
          title="tesla-cat"
        />
      </WithStyles(GridListTile)>
    </WithTheme(Grow)>
    <WithTheme(Grow)
      in={true}
      key="91"
      timeout={1500}
    >
      <WithStyles(GridListTile)
        className="image-library-list__tile"
      >
        <img
          className="image-library-list__tile-img"
          src="https://infocron.s3.us-east-2.amazonaws.com/user-5/nc6DTVcBCpU3Vnew8QuoZkklwvN2uWukO8fPfrSg_thumb450.jpeg"
        />
        <WithStyles(GridListTileBar)
          actionIcon={
            <WithStyles(IconButton)
              onClick={undefined}
            >
              <pure(Delete)
                nativeColor="white"
              />
            </WithStyles(IconButton)>
          }
          title="tesla-cat"
        />
      </WithStyles(GridListTile)>
    </WithTheme(Grow)>
  </WithStyles(GridList)>
</React.Fragment>
`;
